plugins/cameraserver/src/main/java/edu/wpi/first/shuffleboard/plugin/cameraserver/source/CameraServerSource.java

```diff
@@ -0,0 +1,152 @@
+      if (NetworkTableUtils.isDelete(flags) || !(value instanceof String[]) || ((String[]) value).length == 0) {
+        setActive(false);
+      } else {
+        String[] urls = sanitizeUrls((String[]) value);
+        if (EqualityUtils.isDifferent(camera.getUrls(), urls)) {
+          camera.setUrls(urls);
+        }
+        setActive(true);
+      }
+    }), 0xFF);
+  }
+
+  private static String[] sanitizeUrls(String... streams) {
+    return Stream.of(streams)
+        .map(url -> url.replaceFirst("(mjpe?g|ip|usb):", ""))
+        .toArray(String[]::new);
+  }
+
+  public static CameraServerSource forName(String name) {
+    return sources.computeIfAbsent(name, CameraServerSource::new);
+  }
+
+  @Override
+  public SourceType getType() {
+    return CameraServerSourceType.INSTANCE;
+  }
+
+  /**
+   * Continuously reads frames from the MJPEG stream as long as the calling thread has not been interrupted. Because
+   * this is a (nearly) infinite loop, this <strong>cannot</strong> be called from the JavaFX application thread. We
+   * call this in an executor service to run asynchronously.
+   */
+  private void grabForever() {
+    if (Platform.isFxApplicationThread()) {
+      throw new IllegalStateException("This may not run on the FX application thread!");
+    }
+    final Thread thread = Thread.currentThread();
+    while (!thread.isInterrupted()) {
+      boolean success = grabOnceBlocking();
+      if (!success) {
+        // Couldn't grab the frame, wait a bit to try again
+        // This may be caused by a lack of connection (such as the robot is turned off) or various other network errors
+        try {
+          Thread.sleep(1000);
+        } catch (InterruptedException e) {
+          thread.interrupt();
+        }
+      }
+    }
+  }
+
+  /**
+   * Grabs the next frame from the MJPEG stream. This will block until a frame is received or an error occurs.
+   *
+   * @return true if a frame was successfully grabbed, false if an error occurred
+   */
+  private boolean grabOnceBlocking() {
+    long frameTime = videoSink.grabFrameNoTimeout(imageStorage);
+    if (frameTime == 0) {
+      log.warning("Error when grabbing frame from camera '" + getName() + "': " + videoSink.getError());
+      return false;
+    } else {
+      if (getData() == null) {
+        setData(new CameraServerData(getName(), toJavaFxImage(imageStorage)));
+      } else {
+        setData(getData().withImage(toJavaFxImage(imageStorage)));
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Creates a new JavaFX image that contains the same information as an OpenCV image.
+   *
+   * @param mat the OpenCV image that should be converted
+   */
+  private Image toJavaFxImage(Mat mat) {
+    MatOfByte mob = new MatOfByte();
+    Imgcodecs.imencode(".bmp", mat, mob); // bmp for lossless conversion; the stream is already compressed enough!
+    return new Image(new ByteArrayInputStream(mob.toArray()));
```

Comment:
Do you want to free the `mob` before exiting this method? This looks like a potential place for a memory leak.

Reply:
I'll make it a field. That'll let OpenCV do the memory management with `imencode`

Reply:
How will that help?

Reply:
It won't be a memory leak if there's no memory reallocations

Reply:
Fair enough. I really dislike the pattern but you have to do ugly things with code when memory management is involved 🤕